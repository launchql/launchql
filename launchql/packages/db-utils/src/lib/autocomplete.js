import { filter } from 'fuzzy';

export const makeAutocompleteFunctionWithInput = (keys) => (answers, input) => {
  input = input || '';
  return new Promise(function (resolve) {
    setTimeout(function () {
      var fuzzyResult = filter(input, keys);
      resolve(
        fuzzyResult.map(function (el) {
          return el.original;
        })
      );
    }, 25);
  });
};
