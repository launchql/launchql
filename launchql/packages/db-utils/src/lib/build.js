import { listModules, getExtensionsAndModulesChanges } from './modules';
import { dirname, basename, resolve } from 'path';
import { skitchPath as sPath } from './paths';
import { readFileSync } from 'fs';

export const build = async (project) => {
  const skitchPath = await sPath();
  const modules = await listModules();
  const modulesWithChanges = await getExtensionsAndModulesChanges(project);
  const native = [];

  const extensions = Object.keys(modules).reduce((m, key) => {
    const mod = modules[key];
    m[key] = {
      ...mod,
      sql: readFileSync(
        resolve(`${skitchPath}/${mod.path}/sql/${key}--${mod.version}.sql`)
      )
        .toString()
        .split('\n')
        .filter((l, i) => i !== 0)
        .join('\n')
    };
    return m;
  }, {});

  const deps = Object.keys(extensions).reduce((m, k) => {
    m[k] = extensions[k].requires;
    return m;
  }, {});

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(sqlmodule, resolved, unresolved) {
    unresolved.push(sqlmodule);
    let edges = deps[sqlmodule];
    if (!edges) {
      native.push(sqlmodule);
      edges = deps[sqlmodule] = [];
    }
    for (let i = 0; i < edges.length; i++) {
      const dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }
    resolved.push(sqlmodule);
    const index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  const resolved = [];
  const unresolved = [];

  dep_resolve(project, resolved, unresolved);

  const sql = [];

  resolved.forEach((extension) => {
    if (native.includes(extension)) {
      sql.push(`CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`);
    } else {
      sql.push(extensions[extension].sql);
    }
  });
  return sql.join('\n');
};
