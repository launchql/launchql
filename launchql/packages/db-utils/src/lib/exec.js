import { exec as shellExec } from 'shelljs';
import { PGUSER, PGPASSWORD, PGHOST, PGPORT } from '@launchql/db-env';

export const execSync = (cmd, opts) => {
  shellExec(cmd, {
    env: {
      ...process.env,
      PGUSER,
      PGPASSWORD,
      PGHOST,
      PGPORT
    },
    ...opts
  });
};
