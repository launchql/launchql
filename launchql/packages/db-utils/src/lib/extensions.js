import { sqitchPath as path } from './paths';
import { listModules } from './modules';
import { writeFileSync, readFileSync } from 'fs';

export const getAvailableExtensions = async () => {
  let modules = await listModules();
  modules = Object.keys(modules).reduce(
    (m, v) => {
      if (m.includes(v)) return m;
      m.push(v);
      return m;
    },
    [
      'address_standardizer',
      'address_standardizer_data_us',
      'bloom',
      'btree_gin',
      'btree_gist',
      'citext',
      'hstore',
      'intarray',
      'pg_trgm',
      'pgcrypto',
      'plpgsql',
      'plperl',
      'plv8',
      'postgis_tiger_geocoder',
      'postgis_topology',
      'postgis',
      'postgres_fdw',
      'unaccent',
      'uuid-ossp'
    ]
  );
  return modules;
};

export const getExtensionInfo = async (packageDir) => {
  if (!packageDir) {
    packageDir = await path();
  }
  const pkgPath = `${packageDir}/package.json`;
  const pkg = require(pkgPath);
  const extname = await getExtensionName(packageDir);
  const version = pkg.version;
  const Makefile = `${packageDir}/Makefile`;
  const controlFile = `${packageDir}/${extname}.control`;
  const sqlFile = `${extname}--${version}.sql`;
  return {
    extname,
    packageDir,
    version,
    Makefile,
    controlFile,
    sqlFile
  };
};

export const getExtensionName = async (packageDir) => {
  if (!packageDir) {
    packageDir = await path();
  }
  const plan = readFileSync(`${packageDir}/sqitch.plan`)
    .toString()
    .split('\n')
    .map((line) => line.trim())
    .filter((l) => l)
    .filter((l) => /^%project=/.test(l));

  if (!plan.length) {
    throw new Error('no plan name!');
  }

  return plan[0].split('=')[1].trim();
};

export const getInstalledExtensions = async () => {
  const info = await getExtensionInfo();
  let extensions;
  try {
    extensions = readFileSync(info.controlFile)
      .toString()
      .split('\n')
      .find((line) => line.match(/^requires/))
      .split('=')[1]
      .split("'")[1]
      .split(',')
      .map((a) => a.trim());
  } catch (e) {
    throw new Error('missing requires from control files or bad syntax');
  }

  return extensions;
};

export const writeExtensionMakefile = async ({ path, extname, version }) => {
  writeFileSync(
    path,
    `EXTENSION = ${extname}
DATA = sql/${extname}--${version}.sql

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
  `
  );
};

export const writeExtensionControlFile = async ({
  path,
  extname,
  extensions,
  version
}) => {
  writeFileSync(
    path,
    `# ${extname} extension
comment = '${extname} extension'
default_version = '${version}'
module_pathname = '$libdir/${extname}'
requires = '${extensions.join(',')}'
relocatable = false
superuser = false
  `
  );
};

export const writeExtensions = async (extensions) => {
  const { controlFile: path, extname, version } = await getExtensionInfo();
  await writeExtensionControlFile({ path, extname, extensions, version });
};
