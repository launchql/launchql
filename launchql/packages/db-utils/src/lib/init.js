import '@launchql/db-template';
import { promisify } from 'util';
import { exec } from 'child_process';
import { sqitchPath as path, skitchPath as sPath } from './paths';
import {
  writeExtensionMakefile,
  writeExtensionControlFile,
  getExtensionInfo
} from './extensions';
import { dirname, basename } from 'path';
import * as shell from 'shelljs';
import { writeFileSync } from 'fs';
const srcPath = dirname(require.resolve('@launchql/db-template'));
import { makePlan } from './plans';
import { sluggify } from './utils';

const makePackage = ({ name, description, author }) => {
  return {
    name,
    version: '0.0.1',
    description,
    author,
    publishConfig: {
      access: 'restricted'
    },
    scripts: {
      test: 'FAST_TEST=1 launchql-templatedb && jest',
      'test:watch': 'FAST_TEST=1 jest --watch'
    },
    devDependencies: {
      '@babel/cli': '7.10.4',
      '@babel/core': '7.10.4',
      '@babel/plugin-proposal-class-properties': '7.10.4',
      '@babel/plugin-proposal-export-default-from': '7.10.4',
      '@babel/plugin-proposal-object-rest-spread': '7.10.4',
      '@babel/plugin-transform-runtime': '7.10.4',
      '@babel/preset-env': '7.10.4',
      '@babel/runtime': '^7.4.2',
      '@launchql/db-testing': 'latest',
      '@launchql/graphql-testing': 'latest',
      'babel-eslint': '10.1.0',
      'babel-jest': '26.1.0',
      'babel-plugin-import-graphql': '2.7.0',
      'babel-plugin-macros': '2.8.0',
      eslint: '^7.3.1',
      'eslint-config-prettier': '^6.10.0',
      'eslint-plugin-prettier': '^3.1.2',
      'graphql-tag': '2.10.3',
      graphql: '^14.0.2',
      jest: '26.1.0',
      prettier: '2.0.5',
      'regenerator-runtime': '^0.13.2'
    },
    dependencies: {}
  };
};

export const init = async ({
  name,
  description,
  author,
  extensions,
  username,
  scoped
}) => {
  await sPath();
  // init sqitch

  const cur = process.cwd();
  // cur is for enabling us to test with this flag:
  if (process.env.INIT_PATH) {
    process.chdir(process.env.INIT_PATH);
  }

  let pkgname, sqitchname;
  if (scoped) {
    pkgname = `@${username}/${name}`;
    sqitchname = `${username}-${name}`;
  } else {
    pkgname = name;
    sqitchname = name;
  }

  const cmd = ['sqitch', 'init', sqitchname, '--engine', 'pg'].join(' ');
  await promisify(exec)(cmd.trim());

  // now we have a sqitch path!

  const sqitchPath = await path();
  const pkg = makePackage({ name: pkgname, description, author });

  // initialize template
  shell.cp('-r', `${srcPath}/sqitch/*`, `${sqitchPath}/`);
  shell.cp('-r', `${srcPath}/sqitch/.*`, `${sqitchPath}/`);

  writeFileSync(`${sqitchPath}/package.json`, JSON.stringify(pkg, null, 2));

  shell.mkdir('-p', `${sqitchPath}/sql`);
  const extname = sluggify(sqitchname);

  const info = await getExtensionInfo();

  await writeExtensionMakefile({
    path: info.Makefile,
    extname,
    version: '0.0.1'
  });
  await writeExtensionControlFile({
    path: info.controlFile,
    extname,
    version: '0.0.1',
    extensions
  });

  const settings = {
    name: sqitchname,
    projects: true
  };

  const plan = await makePlan(sqitchPath, settings);

  writeFileSync(`${sqitchPath}/sqitch.plan`, plan);
  writeFileSync(
    `${sqitchPath}/.npmignore`,
    `# NOTE keeping this minimal since we generally want everything

*.log
npm-debug.log*

# Dependency directories
node_modules

# npm package lock
package-lock.json
yarn.lock    
`
  );
  writeFileSync(
    `${sqitchPath}/.gitignore`,
    `node_modules
`
  );

  process.chdir(cur);
};

export const initSkitch = async () => {
  const dir = process.cwd();
  shell.cp('-r', `${srcPath}/template/*`, `${dir}/`);
  shell.cp('-r', `${srcPath}/template/.*`, `${dir}/`);
  const name = sluggify(basename(process.cwd()));
  const pkg = {
    private: true,
    name,
    scripts: {
      build: 'lerna run prepare --parallel',
      bootstrap: 'lerna bootstrap --use-workspaces'
    },
    devDependencies: {
      '@babel/cli': '7.12.1',
      '@babel/core': '7.12.3',
      '@pyramation/babel-preset-env': '0.1.0',
      'babel-eslint': '10.1.0',
      'babel-jest': '26.6.1',
      eslint: '7.12.1',
      'eslint-config-prettier': '^6.10.0',
      'eslint-plugin-prettier': '^3.1.2',
      jest: '26.6.1',
      lerna: '3.22.1',
      prettier: '2.1.2'
    },
    workspaces: ['packages/*']
  };
  writeFileSync(`${process.cwd()}/package.json`, JSON.stringify(pkg, null, 2));
  const lerna = {
    lerna: '3.4.2',
    useWorkspaces: true,
    npmClient: 'yarn',
    npmClientArgs: ['--no-lockfile'],
    packages: ['packages/*'],
    version: 'independent',
    registry: 'https://registry.npmjs.org',
    command: {
      create: {
        license: 'SEE LICENSE IN LICENSE',
        access: 'restricted'
      },
      publish: {
        allowBranch: 'master'
      }
    }
  };
  writeFileSync(`${process.cwd()}/lerna.json`, JSON.stringify(lerna, null, 2));
  const ignore = `node_modules
.DS_Store
.eslintcache
*.log
**/node_modules
coverage
packages/**/build
packages/**/main
packages/**/module`;
  writeFileSync(`${process.cwd()}/.gitignore`, ignore);
  writeFileSync(`${process.cwd()}/.npmignore`, ignore);
  writeFileSync(
    `${process.cwd()}/Makefile`,
    `
up:
\tdocker-compose up -d

down:
\tdocker-compose down -v

ssh:
\tdocker exec -it ${name}-postgres /bin/bash

install:
\tdocker exec ${name}-postgres /sql-bin/install.sh

  `
  );
  writeFileSync(
    `${process.cwd()}/docker-compose.yml`,
    `version: "2"
services:
  postgres:
    container_name: ${name}-postgres
    image: pyramation/postgis
    environment:
      - "POSTGRES_USER=postgres"
      - "POSTGRES_PASSWORD=password"
    ports:
      - "5432:5432"
    expose:
      - "5432"
    volumes:
      - ./bin:/sql-bin
      - ./packages:/sql-packages
      - ./extensions:/sql-extensions
  `
  );
};
