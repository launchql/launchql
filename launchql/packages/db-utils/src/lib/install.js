import { resolve, join, dirname, basename } from 'path';
import { sync as mkdirp } from 'mkdirp';
import { sync as glob } from 'glob';
import { sync as rimraf } from 'rimraf';
import { skitchPath } from './paths';
import { exec } from 'shelljs';
import parse from 'parse-package-name';

const TMPDIR = process.env.TMPDIR;
const rnd = () =>
  Math.random().toString(36).substring(2, 15) +
  Math.random().toString(36).substring(2, 15);

export const installPackage = async (pkgstr) => {
  const { name, version } = parse(pkgstr);

  const sPath = await skitchPath();
  const cur = process.cwd();

  /*

currently creating a temp dir and treating this as a global skitch ext that any sqitch proj can use...

IMPROVEMENTS:

later maybe you can install in the current dir, and then install everything so that it's actually in the current dir, and then

1. updates the package.json of the sqitch project
2. also the control file of the sqitch project

(leaves skitch project alone, only updates the individual projects)

Questions: how about if you install one that already exists? Do we just leave it there or delete and update?

Note: What about when you create a new sqitch (e.g. lql init) and you choose modules... notice it doesn't yet update the package.json???

*/

  const tmp = join(TMPDIR, rnd());
  mkdirp(tmp);
  process.chdir(tmp);
  exec(`npm install ${name}@${version} --production --prefix ./extensions`);
  const sqitch = glob('./extensions/**/sqitch.conf');
  const cmds = sqitch
    .map((f) => resolve(join(tmp, f)))
    .map((conf) => {
      const extDir = dirname(conf);
      const dir = extDir.split('node_modules/')[1];
      return [extDir, resolve(join(sPath, 'extensions', dir)), dir];
    });

  for (const [src, dst, pkg] of cmds) {
    // sorry not sorry
    rimraf(dst);
    console.log(`installing ${pkg}...`);
    mkdirp(dirname(dst));
    exec(`mv ${src} ${dst}`);
  }

  rimraf(tmp);

  process.chdir(cur);

  console.log(`
  WARNING!
  
  manually add this to package.json to the modules that need it if you plan to publish them
  `);
};
