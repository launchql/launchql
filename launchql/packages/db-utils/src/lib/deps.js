import { readFileSync } from 'fs';
import { relative } from 'path';
import { sync as glob } from 'glob';
import { listModules } from './modules';
import { getExtensionName } from './extensions';

const makeKey = (sqlmodule) => '/deploy/' + sqlmodule + '.sql';

export const getDeps = async (packageDir) => {
  const extname = await getExtensionName(packageDir);

  var external = [];

  var deps = {};

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(sqlmodule, resolved, unresolved) {
    unresolved.push(sqlmodule);
    let edges = deps[makeKey(sqlmodule)];
    if (!edges) {
      if (/:/.test(sqlmodule)) {
        external.push(sqlmodule);
        edges = deps[sqlmodule] = [];
      } else {
        throw new Error(`no module ${sqlmodule}`);
      }
    }

    for (var i = 0; i < edges.length; i++) {
      var dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(sqlmodule);
    var index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  var files = await glob(`${packageDir}/deploy/**/**.sql`);

  for (var i = 0; i < files.length; i++) {
    const data = readFileSync(files[i]);
    var lines = data.toString().split('\n');
    const key = '/' + relative(packageDir, files[i]);
    deps[key] = [];

    for (var j = 0; j < lines.length; j++) {
      var m = lines[j].match(/^-- requires: (.*)/);
      if (m) {
        deps[key].push(m[1].trim());
      }

      // check only:
      var m2;
      var keyToTest;

      if (/:/.test(lines[j])) {
        m2 = lines[j].match(/^-- Deploy ([^:]*):([\w\/]+) to pg/);
        if (m2) {
          keyToTest = m2[2];
          if (extname !== m2[1]) {
            throw new Error(
              'referencing bad project name inside of deploy file\n' + lines[j]
            );
          }
          if (key != makeKey(keyToTest)) {
            throw new Error(
              'deployment script in wrong place or is named wrong internally\n' +
                lines[j]
            );
          }
        }
      } else {
        m2 = lines[j].match(/^-- Deploy (.*) to pg/);
        if (m2) {
          keyToTest = m2[1];
          if (key != makeKey(keyToTest)) {
            throw new Error(
              'deployment script in wrong place or is named wrong internally\n' +
                lines[j]
            );
          }
        }
      }
    }
  }

  let resolved = [];
  var unresolved = [];

  deps = Object.assign(
    {
      [makeKey('apps/index')]: Object.keys(deps)
        .filter((dep) => dep.match(/^\/deploy\//))
        .map((dep) => dep.replace(/^\/deploy\//, '').replace(/.sql$/, ''))
    },
    deps
  );

  dep_resolve('apps/index', resolved, unresolved);
  var index = resolved.indexOf('apps/index');
  resolved.splice(index);
  delete deps[makeKey('apps/index')];

  // move extensions up
  // TODO maybe deprecate this?
  const extensions = resolved.filter((a) => a.match(/^extensions/));
  const normalSql = resolved.filter((a) => !a.match(/^extensions/));
  // console.log({ extensions, normalSql });
  // resolved = useExtensions ? [...extensions, ...normalSql] : [...normalSql];
  resolved = [...extensions, ...normalSql];
  // resolved = [...normalSql];

  return {
    external,
    resolved,
    deps
  };
};

export const extDeps = async (name) => {
  const modules = await listModules();
  const external = [];
  if (!modules[name]) {
    throw new Error(`module ${name} does not exist!`);
  }
  // build deps
  const deps = Object.keys(modules).reduce((memo, key) => {
    memo[key] = modules[key].requires;
    return memo;
  }, {});

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(sqlmodule, resolved, unresolved) {
    unresolved.push(sqlmodule);
    let edges = deps[sqlmodule];
    if (!edges) {
      external.push(sqlmodule);
      edges = deps[sqlmodule] = [];
    }

    for (var i = 0; i < edges.length; i++) {
      var dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(sqlmodule);
    var index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  const resolved = [];
  var unresolved = [];

  dep_resolve(name, resolved, unresolved);

  return { external, resolved };
};
