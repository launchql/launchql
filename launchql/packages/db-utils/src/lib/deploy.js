import * as shell from 'shelljs';
import { resolve } from 'path';

import { listModules } from './modules';

import { skitchPath } from './paths';
import { extDeps } from './deps';

import { PGUSER, PGPASSWORD, PGHOST, PGPORT, PATH } from '@launchql/db-env';

const pg = require('pg');

// should we be parsing the plan file?
// currently assuming only extensions in control file...

export const deploy = async (name, database, opts) => {
  const modules = await listModules();
  const path = await skitchPath();
  if (!modules[name]) {
    throw new Error(`module ${name} does not exist!`);
  }
  const extensions = await extDeps(name);

  const pgPool = new pg.Pool({
    connectionString: `postgres://${PGUSER}:${PGPASSWORD}@${PGHOST}:${PGPORT}/${database}`
  });

  for (let i = 0; i < extensions.resolved.length; i++) {
    const extension = extensions.resolved[i];
    try {
      if (extensions.external.includes(extension)) {
        console.log(`CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`);
        await pgPool.query(
          `CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`
        );
      } else {
        console.log(modules[extension].path);
        console.log(`sqitch deploy db:pg:${database}`);
        const cmd = shell.exec(`sqitch deploy db:pg:${database}`, {
          cwd: resolve(path, modules[extension].path),
          env: {
            PGUSER,
            PGPASSWORD,
            PGHOST,
            PGPORT,
            PATH
          }
        });
        if (cmd.code !== 0) {
          throw new Error('deploy failed');
        }
      }
    } catch (e) {
      pgPool.end();
      process.exit(1);
    }
  }
  pgPool.end();

  return extensions;
};
