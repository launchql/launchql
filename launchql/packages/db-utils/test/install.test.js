import { install, installPackage } from '../src/lib/install';
import { init, initSkitch } from '../src/lib/init';
import { sync as mkdirp } from 'mkdirp';
import { sync as glob } from 'glob';
import { sync as rimraf } from 'rimraf';
import { writeFileSync } from 'fs';

const TMPDIR = process.env.TMPDIR;
const rnd = () =>
  Math.random().toString(36).substring(2, 15) +
  Math.random().toString(36).substring(2, 15);

describe('installs', () => {
  let dir, projDir;
  beforeEach(async () => {
    dir = TMPDIR + '/' + rnd();
    projDir = dir + '/myproject';
    mkdirp(projDir);
    process.chdir(dir);
    await initSkitch();
    process.chdir(projDir);
    await init({
      name: 'myproject',
      description: 'my amazing project',
      author: 'dan@example.com',
      extensions: ['plpgsql', 'citext']
    });
  });
  afterEach(() => {
    rimraf(dir);
  });
  describe('skitch path', () => {
    it('sqitch install', async () => {
      const files = glob('**');
      expect(files).toMatchSnapshot();
      process.chdir(dir);
      writeFileSync(
        `${dir}/package.json`,
        JSON.stringify(
          {
            name: 'blankpackage'
          },
          null,
          2
        )
      );
      await installPackage('skitch-extension-verify', 'latest');
      const filesAfter = glob('**');
      expect(filesAfter).toMatchSnapshot();
    });
  });
});
