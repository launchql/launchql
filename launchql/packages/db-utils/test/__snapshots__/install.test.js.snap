// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`installs skitch path sqitch install 1`] = `
Array [
  "deploy",
  "Makefile",
  "myproject.control",
  "package.json",
  "revert",
  "sqitch.conf",
  "sqitch.plan",
  "sql",
  "test",
  "test/first.gql.test.js",
  "test/first.sql.test.js",
  "test/utils",
  "test/utils/graphql.js",
  "test/utils/index.js",
  "verify",
]
`;

exports[`installs skitch path sqitch install 2`] = `
Array [
  "babel.config.js",
  "bin",
  "bin/install.sh",
  "bootstrap-roles.sql",
  "docker-compose.yml",
  "extensions",
  "extensions/skitch-extension-utils",
  "extensions/skitch-extension-utils/deploy",
  "extensions/skitch-extension-utils/deploy/procedures",
  "extensions/skitch-extension-utils/deploy/procedures/get_entity_from_str.sql",
  "extensions/skitch-extension-utils/deploy/procedures/get_schema_from_str.sql",
  "extensions/skitch-extension-utils/deploy/procedures/list_indexes.sql",
  "extensions/skitch-extension-utils/deploy/procedures/list_memberships.sql",
  "extensions/skitch-extension-utils/deploy/procedures/tg_update_timestamps.sql",
  "extensions/skitch-extension-utils/LICENSE",
  "extensions/skitch-extension-utils/Makefile",
  "extensions/skitch-extension-utils/package.json",
  "extensions/skitch-extension-utils/revert",
  "extensions/skitch-extension-utils/revert/procedures",
  "extensions/skitch-extension-utils/revert/procedures/get_entity_from_str.sql",
  "extensions/skitch-extension-utils/revert/procedures/get_schema_from_str.sql",
  "extensions/skitch-extension-utils/revert/procedures/list_indexes.sql",
  "extensions/skitch-extension-utils/revert/procedures/list_memberships.sql",
  "extensions/skitch-extension-utils/revert/procedures/tg_update_timestamps.sql",
  "extensions/skitch-extension-utils/skitch-extension-utils.control",
  "extensions/skitch-extension-utils/sqitch.conf",
  "extensions/skitch-extension-utils/sqitch.plan",
  "extensions/skitch-extension-utils/sql",
  "extensions/skitch-extension-utils/sql/skitch-extension-utils--0.0.7.sql",
  "extensions/skitch-extension-utils/verify",
  "extensions/skitch-extension-utils/verify/procedures",
  "extensions/skitch-extension-utils/verify/procedures/get_entity_from_str.sql",
  "extensions/skitch-extension-utils/verify/procedures/get_schema_from_str.sql",
  "extensions/skitch-extension-utils/verify/procedures/list_indexes.sql",
  "extensions/skitch-extension-utils/verify/procedures/list_memberships.sql",
  "extensions/skitch-extension-utils/verify/procedures/tg_update_timestamps.sql",
  "extensions/skitch-extension-verify",
  "extensions/skitch-extension-verify/deploy",
  "extensions/skitch-extension-verify/deploy/procedures",
  "extensions/skitch-extension-verify/deploy/procedures/verify_constraint.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_domain.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_extension.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_function.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_index.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_membership.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_policy.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_role.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_schema.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_security.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_table_grant.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_table.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_trigger.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_type.sql",
  "extensions/skitch-extension-verify/deploy/procedures/verify_view.sql",
  "extensions/skitch-extension-verify/LICENSE",
  "extensions/skitch-extension-verify/Makefile",
  "extensions/skitch-extension-verify/package.json",
  "extensions/skitch-extension-verify/revert",
  "extensions/skitch-extension-verify/revert/procedures",
  "extensions/skitch-extension-verify/revert/procedures/verify_constraint.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_domain.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_extension.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_function.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_index.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_membership.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_policy.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_role.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_schema.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_security.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_table_grant.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_table.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_trigger.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_type.sql",
  "extensions/skitch-extension-verify/revert/procedures/verify_view.sql",
  "extensions/skitch-extension-verify/skitch-extension-verify.control",
  "extensions/skitch-extension-verify/sqitch.conf",
  "extensions/skitch-extension-verify/sqitch.plan",
  "extensions/skitch-extension-verify/sql",
  "extensions/skitch-extension-verify/sql/skitch-extension-verify--0.0.7.sql",
  "extensions/skitch-extension-verify/verify",
  "extensions/skitch-extension-verify/verify/procedures",
  "extensions/skitch-extension-verify/verify/procedures/verify_constraint.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_domain.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_extension.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_function.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_index.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_membership.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_policy.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_role.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_schema.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_security.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_table_grant.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_table.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_trigger.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_type.sql",
  "extensions/skitch-extension-verify/verify/procedures/verify_view.sql",
  "jest.config.js",
  "lerna.json",
  "Makefile",
  "myproject",
  "myproject/deploy",
  "myproject/Makefile",
  "myproject/myproject.control",
  "myproject/package.json",
  "myproject/revert",
  "myproject/sqitch.conf",
  "myproject/sqitch.plan",
  "myproject/sql",
  "myproject/test",
  "myproject/test/first.gql.test.js",
  "myproject/test/first.sql.test.js",
  "myproject/test/utils",
  "myproject/test/utils/graphql.js",
  "myproject/test/utils/index.js",
  "myproject/verify",
  "package.json",
  "readme.md",
  "skitch.json",
]
`;
